services/headless-lms/server/src/controllers/course_material/exercises.rs

```diff
@@ -96,6 +98,15 @@ async fn post_submission(
     let mut conn = pool.acquire().await?;
     let exercise = models::exercises::get_by_id(&mut conn, *exercise_id).await?;
 
+    // Checking if exercise has a deadline and if so whether it is passed
+    if let Some(deadline) = exercise.deadline {
+        if Utc::now() >= deadline {
```

Comment:
1 second grace time to account clock drift.

Take the deadline by default from chapter and let exercise.deadline to override it.